src/Generics/Deriving/Default.hs

```diff
@@ -0,0 +1,251 @@
+-- it is now possible to choose from many
+-- default instantiations.
+--
+-- This package contains a number of such classes. This module demonstrates
+-- how one might create a family of newtypes ('Default', 'Default1') for
+-- which such instances are defined.
+--
+-- One might then use
+-- @<https://downloads.haskell.org/~ghc/8.6.3/docs/html/users_guide/glasgow_exts.html?highlight=derivingvia#extension-DerivingVia DerivingVia>@
+-- as follows. The implementations of the data types are elided here (they
+-- are irrelevant). For most cases, either the deriving clause with the
+-- data type definition or the standalone clause will work (for some types
+-- it is necessary to supply the context explicitly using the latter form).
+-- See the source of this module for the implementations of instances for
+-- the 'Default' family of newtypes and the source of the test suite for
+-- some types which derive instances via these wrappers.
+
+{-# LANGUAGE CPP #-}
+#if __GLASGOW_HASKELL__ >= 701
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE Safe #-}
+#endif
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Generics.Deriving.Default
+  ( -- * Kind Type
+
+    -- $default
+
+    Default(..)
+
+  , -- * Kind (Type -> Type)
+
+    -- $default1
+
+    Default1(..)
+
+    -- * Other kinds
+
+    -- $other-kinds
+  ) where
+
+#if !(MIN_VERSION_base(4,8,0))
+import Control.Applicative ((<$>))
+#endif
+import Control.Monad (liftM)
+
+import Generics.Deriving.Base
+import Generics.Deriving.Copoint
+import Generics.Deriving.Enum
+import Generics.Deriving.Eq
+import Generics.Deriving.Foldable
+import Generics.Deriving.Functor
+import Generics.Deriving.Monoid
+import Generics.Deriving.Semigroup
+import Generics.Deriving.Show
+import Generics.Deriving.Traversable
+import Generics.Deriving.Uniplate
+
+-- $default
+--
+-- For classes which take an argument of kind 'Data.Kind.Type', use 'Default'.
+--
+-- These examples use 'GShow' and 'GEq'; they are interchangeable.
+--
+-- @
+-- data MyType = …
+--  deriving (Generic)
+--  deriving (GEq) via (Default MyType)
+--
+-- deriving via (Default MyType) instance GShow MyType
+-- @
+--
+-- Instances may be parameterized by type variables.
+--
+-- @
+-- data MyType1 a = …
+--  deriving (Generic)
+--  deriving (GShow) via (Default (MyType1 a))
+--
+-- deriving via Default (MyType1 a) instance GEq a => GEq (MyType1 a)
+-- @
+
+-- | This newtype wrapper can be used to derive default instances for
+-- classes taking an argument of kind 'Data.Kind.Type'.
+newtype Default a = Default { unDefault :: a }
+
+-- $default1
+--
+-- For classes which take an argument of kind @'Data.Kind.Type' -> 'Data.Kind.Type'@, use 'Default1'.
+--
+-- Unlike for 'MyType1', there can be no implementation of these classes for @MyType :: 'Data.Kind.Type'@.
+--
+-- @
+-- data MyType1 a = …
+--  deriving (Generic)
```

Comment:
Apparently not.

```haskell
data MyType1 a = MyType1 a
  deriving (Generic1)
  deriving (GShow) via (Default (MyType1 a))

deriving via Default (MyType1 a) instance GEq a => GEq (MyType1 a)
```

```
    • No instance for (Generic (MyType1 a))
        arising from the 'deriving' clause of a data type declaration
      Possible fix:
        use a standalone 'deriving instance' declaration,
          so you can specify the instance context yourself
    • When deriving the instance for (GShow (MyType1 a))
    |
130 |   deriving (GShow) via (Default (MyType1 a))
    |
```

I ought to add this to the documentation — presumably it is because `gshowsPrecdefault` (which the `Default` instance uses) has a `Generic a` constraint? I get an equivalent error for `GEq` as `GShow`.

https://github.com/dreixel/generic-deriving/blob/413733b81db99c2bd8593c315c912d4688ecb7f7/src/Generics/Deriving/Show.hs#L198-L200

Reply:
Oops, I totally missed the fact that this is using `Default`, not `Default1`. Please ignore my suggestion.

Reply:
Nope, you're right — I used the wrong `MyType1` example!